nntrainer/tensor/tensor.cpp

```diff
@@ -3041,17 +3041,46 @@ void Tensor::copyData(const Tensor &from) {
   if (size() != from.size())
     throw std::invalid_argument("Size of tensor to copy must match");
 
-  if (getDataType() != from.getDataType())
-    throw std::invalid_argument("Data type of tensor to copy must match");
-
-  if (getDataType() == ml::train::TensorDim::DataType::FP32) {
-    copy(from.getData<float>());
-  } else if (getDataType() == ml::train::TensorDim::DataType::FP16) {
+  if (getDataType() == from.getDataType()) {
+    if (getDataType() == ml::train::TensorDim::DataType::FP32) {
+      copy(from.getData<float>());
+    } else if (getDataType() == ml::train::TensorDim::DataType::FP16) {
 #ifdef ENABLE_FP16
-    copy(from.getData<_FP16>());
+      copy(from.getData<_FP16>());
 #else
-    throw std::invalid_argument("Error: enable-fp16 is not enabled");
+      throw std::invalid_argument("Error: enable-fp16 is not enabled");
+#endif
+    } else {
+      copy(from.getData<uint8_t>());
+    }
+  } else {
+    if (getDataType() == ml::train::TensorDim::DataType::FP32) {
+      if (from.getDataType() == ml::train::TensorDim::DataType::FP16) {
+#ifdef ENABLE_FP16
+        scopy(size(), from.getData<_FP16>(), 1, getData<float>(), 1);
+#else
+        throw std::invalid_argument("Error: enable-fp16 is not enabled");
 #endif
+      } else if (from.getDataType() == ml::train::TensorDim::DataType::QINT8) {
+        scopy_int8_to_float32(from.size(), from.getData<uint8_t>(), 1,
+                              getData<float>(), 1);
+      } else if (from.getDataType() == ml::train::TensorDim::DataType::QINT4) {
+        scopy_int4_to_float32((from.size() + 1) / 2, from.getData<uint8_t>(), 1,
+                              getData<float>(), 1);
+      }
+    } else if (getDataType() == ml::train::TensorDim::DataType::FP16) {
+#ifdef ENABLE_FP16
+      if (from.getDataType() == ml::train::TensorDim::DataType::QINT8) {
+        scopy_int8_to_float16(from.size(), from.getData<uint8_t>(), 1,
```

Comment:
Quick question:
I am considering renaming the term `scopy` to something else...
"s" here means "single precision(fp32)", so it really cannot be used in situation like this.
Any idea for the name of copying function of int8 <-> int4 <-> float16?

Reply:
I don't think there are standard prefixes for these data types. maybe we could remove 's' and go with `copy_int8_float16`.

Reply:
Think so too.
Thanks!